File: iosApp/iosApp/Modules/ImprovementProposals ✅/ImprovementProposalsWireframe.swift
// Created by web3d3v on 30/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultImprovementProposalsWireframe {
    private weak var parent: UIViewController?
    private let improvementProposalWireframeFactory: ImprovementProposalWireframeFactory
    private let improvementProposalsService: ImprovementProposalsService
    private weak var vc: UIViewController?

    init(
        _ parent: UIViewController?,
        improvementProposalWireframeFactory: ImprovementProposalWireframeFactory,
        improvementProposalsService: ImprovementProposalsService
    ) {
        self.parent = parent
        self.improvementProposalWireframeFactory = improvementProposalWireframeFactory
        self.improvementProposalsService = improvementProposalsService
    }
}

extension DefaultImprovementProposalsWireframe: ImprovementProposalsWireframe {

    func present() {
        let vc = wireUp()
        self.vc = vc
        parent?.show(vc, sender: self)
    }

    func navigate(destination_______ destination: ImprovementProposalsWireframeDestination) {
        if let vote = destination as? ImprovementProposalsWireframeDestination.Vote {
            guard let url = voteUrl(vote.proposal) else { return }
            UIApplication.shared.open(url)
        }
        if let dest = destination as? ImprovementProposalsWireframeDestination.Proposal {
            improvementProposalWireframeFactory.make(
                vc,
                context: .init(proposal: dest.proposal)
            ).present()
        }
        if (destination as? ImprovementProposalsWireframeDestination.Dismiss) != nil {
            vc?.popOrDismiss()
        }
    }
}

extension DefaultImprovementProposalsWireframe {

    func wireUp() -> UIViewController {
        let interactor = DefaultImprovementProposalsInteractor(
            improvementProposalsService: improvementProposalsService
        )
        let storyboard = UIStoryboard(.improvementProposals)
        let vc: ImprovementProposalsViewController = storyboard.instantiate()
        let presenter = DefaultImprovementProposalsPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor
        )
        vc.presenter = presenter
        return NavigationController(rootViewController: vc)
    }

    func voteUrl(_ proposal: ImprovementProposal) -> URL? {
        let body = proposal.tweet.addingPercentEncoding(
            withAllowedCharacters: .urlHostAllowed
        ) ?? ""
        let url = proposal.pageUrl.addingPercentEncoding(
            withAllowedCharacters: .urlHostAllowed
        ) ?? ""
        let str = "https://www.twitter.com/intent/tweet?text=\(body)&url=\(url)"
        return URL(string: str)
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/improvementproposals/ImprovementProposalsWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.improvementproposals

import android.content.Intent
import android.net.Uri
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.appContext
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3wallet.android.modules.improvementproposal.ImprovementProposalWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.improvementProposal.ImprovementProposalWireframeContext
import com.sonsofcrypto.web3walletcore.modules.improvementProposals.DefaultImprovementProposalsInteractor
import com.sonsofcrypto.web3walletcore.modules.improvementProposals.DefaultImprovementProposalsPresenter
import com.sonsofcrypto.web3walletcore.modules.improvementProposals.ImprovementProposalsWireframe
import com.sonsofcrypto.web3walletcore.modules.improvementProposals.ImprovementProposalsWireframeDestination
import com.sonsofcrypto.web3walletcore.services.improvementProposals.ImprovementProposalsService

class DefaultImprovementProposalsWireframe(
    private val parent: WeakRef<Fragment>?,
    private val improvementProposalsService: ImprovementProposalsService,
    private val improvementProposalWireframeFactory: ImprovementProposalWireframeFactory,
): ImprovementProposalsWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.push(fragment, true)
    }

    override fun navigate(destination: ImprovementProposalsWireframeDestination) {
        when (destination) {
            is ImprovementProposalsWireframeDestination.Proposal -> {
                val context = ImprovementProposalWireframeContext(destination.proposal)
                improvementProposalWireframeFactory.make(parent?.get(), context).present()
            }
            is ImprovementProposalsWireframeDestination.Vote -> {
                val body = Uri.encode(destination.proposal.tweet)
                val url = Uri.encode(destination.proposal.pageUrl)
                val str = "https://www.twitter.com/intent/tweet?text=$body&url=$url"
                val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(str))
                ContextCompat.startActivity(appContext, browserIntent, null)
            }
            is ImprovementProposalsWireframeDestination.Dismiss -> {
                parent?.navigationFragment?.popOrDismiss()
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = ImprovementProposalsFragment()
        val interactor = DefaultImprovementProposalsInteractor(
            improvementProposalsService
        )
        val presenter = DefaultImprovementProposalsPresenter(
            WeakRef(view),
            this,
            interactor
        )
        view.presenter = presenter
        return view
    }
}